### ISSUE TITLE
SocketException "Unknown error (0xffffffff)" at System.Net.Http.ConnectHelper.ConnectAsync

### ISSUE THREAD

**Issue by kierenj** [dotnet/runtime#28736]
Type name: `System.Net.Sockets.SocketException`
`.HResult` = -2147467259
`.Message` = "Unknown error (0xffffffff)"
`.NativeErrorCode` = -1
`.Source` = "System.Private.CoreLib"

The stack trace for the outer exception:

```
at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
```

We started receiving this error a few hours ago (after weeks / months of no problem), and a process restart was required to fix it.  It occurred on every outbound HTTP request until the process was killed.  Process memory use was not particularly high.

Azure App Service.  Kudu versioning:
![image](https://user-images.githubusercontent.com/82343/53102301-4b99aa00-3523-11e9-8455-865c97911c44.png)

Other versions:
```
my\bin\folder> dotnet --version
2.2.102
```

From csproj:
```
<TargetFramework>netcoreapp2.1</TargetFramework>
```

I checked the TCP Connections diagnostics tool on Azure - there were only ~1000 connections being used (so I guess not port exhaustion?).

Interestingly - we're using StackExchange.Redis and at the time of the first exception, I do see that there was a connection error with Redis too.

What can I do to assist?  The problem occurred in production, and while I have app-level logging, but not too much else.

Edit: the issue just re-occurred on the same server, after the process restart

**Comment by karelz** [comment]
I am afraid there is much we can do at this moment. I think we would need someone able to try it out on .NET Core 3.1 and possibly on custom .NET 5 builds (self-contained) with additional logging to discover what is going on. Nothing else comes to mind.
Closing for now as suggested above -- we have 2 people hitting it. If anyone is in position to experiment, please let us know and we can try to get to the bottom of this ...

**Comment by buddalasunil999** [comment]
I am having the same issue. I am running a .net core app as background service in Ubuntu server. Please check the stack trace below.

Unknown error -1) ---> System.Net.Http.HttpRequestException: Unknown error -1 ---> System.Net.Sockets.SocketException: Unknown error -1 at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken) --- End of inner exception stack trace --- at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken) at System.Threading.Tasks.ValueTask`1.get_Result() at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken) at System.Threading.Tasks.ValueTask`1.get_Result() at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask) at System.Threading.Tasks.ValueTask`1.get_Result() at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)

**Comment by wfurt** [comment]
What version are you on @buddalasunil999? If you can reproduce it, can you check the OS call with `strace` and get the errno?

**Comment by buddalasunil999** [comment]
@wfurt I am using .net core 2.1

**Comment by wfurt** [comment]
that may be https://github.com/dotnet/runtime/issues/28630 fixed in 3.0. strace would tell for sure. You should give it try with 3.1

**Comment by kierenj** [comment]
The original (Windows) issue would seem to be unrelated unfortunately. We have another project showing the error now, but on this project its interspersed some SQL exceptions with a different win32 exception:

System.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 0 - An operation on a socket could not be performed because the system lacked sufficient buffer space or because a queue was full.) ---> System.ComponentModel.Win32Exception (10055): An operation on a socket could not be performed because the system lacked sufficient buffer space or because a queue was full
   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)

(Revelant bit: System.ComponentModel.Win32Exception (10055): An operation on a socket could not be performed because the system lacked sufficient buffer space or because a queue was full)

I don't know if this could be a clue ..?

**Comment by rogatec** [comment]
I had this same issue today and the google search brought me here.

#### Situation
On my workstation I have an ActiveMQ ArtemisBroker running (just http://localhost:5672) and was developing a `.net core 3.1` cli app with MassTransit.ActiveMQ library.
After work and back at home I continued work and suddenly the .net core app could not connect to the Broker anymore.
I got this same weird `System.Net.Sockets.SocketException with "Unknown error (0xffffffff)" in "System.Private.CoreLib"`

I fixed it by closing my vpn connection (which I had running for other work stuff on the network drive..) and restartet the .net core app.

Perhaps this can also be helpful information.

**Comment by wfurt** [comment]
If this happens to anybody it would be great to set breakpoint on ConnectEx and check return code from the call. That would allow up to triage if this is something wrong at .NET layer or this is caused by some underlying OS condition.
At the same time, I would suggest to check TCP table and maybe system/security logs.

**Comment by oakad** [comment]
.NET Core SDK (3.1.201)

I'm hitting this with all new code (AWS head/upload) on Centos around 6 times out of 10 - the problem is pretty persistent.

**Comment by oakad** [comment]
It seems that the issue is strongly related to the rate of task creation. Also, it appears that runtime in general has some bugginess and can get stuck outright, spinning on cpu, when large number of tasks are created in rapid succession.

I have implemented 2 relatively simple programs in go and c#, both uploading large number of files to s3 using corresponding SDKs (one upload per task/goroutine). Go can easily max out the S3 service, hitting the service RPS limit. Dotnet Core can get stuck. :-) 

Things appear to work ok when I introduce semaphore based limiters here and there.

* It is possible that I'm doing something very wrong. However, I reviewed my code several times and there appears to be certain threshold of number of simultaneous tasks which breaks things.

**Comment by stephentoub** [comment]
@oakad, repro? And why do you think it's related to this closed issue?

**Comment by oakad** [comment]
Because it's exactly the error I was getting. Same location, same message.

**Comment by stephentoub** [comment]
Can you share your repro? I'm confused by the description of the problem: this issue is about an exception, but I interpret your description to be about a hang ("gets stuck").

**Comment by oakad** [comment]
Here's a stack trace easily - version .NET Core SDK (3.1.201) as I said.

As to repro it's a bit tricky.

Exceptions were encountered: System.AggregateException: One or more errors occurred. (Bad value for ai_flags)
 ---> System.Net.Http.HttpRequestException: Bad value for ai_flags
 ---> System.Net.Sockets.SocketException (0xFFFFFFFF): Bad value for ai_flags
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Amazon.Runtime.HttpWebRequestMessage.GetResponseAsync(CancellationToken cancellationToken)
   at Amazon.Runtime.Internal.HttpHandler`1.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RedirectHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.Unmarshaller.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.S3.Internal.AmazonS3ResponseHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CredentialsRetriever.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.S3.Internal.AmazonS3ExceptionHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.MetricsHandler.InvokeAsync[T](IExecutionContext executionContext)

**Comment by oakad** [comment]
My other comment was somewhat unrelated. If we look at the issue at hand, that is, socket error, it appears to be strongly bound to the rate, at which those AWS SDK async tasks are created.

My other problem was apparently due to memory pressure - it was connected to OS OOM handler being invoked (but no internal OOM exception).

**Comment by kshaa** [comment]
I also encountered a similiar problem today both on .NET Core 3.1 and .NET Core 5.0 Preview 7.0 while running Ubuntu 20.04 LTS.

I was sending an HTTP GET request to `http://www.example.com/Pages/Home.aspx` (not actually example.com) and received back a 302 status code with `Location: https://www.example.comPages/Home.aspx`, which is obviously a bug in the endpoint I was communicating with. However the `HttpWebRequest` class that I used for executing HTTP requests had `AllowAutoRedirect = true` set, so it tried to follow the redirect. That either resulted in the low level syscall `getaddrinfo` to fail or something to that end. Whatever the low-level reason was, it resulted in .NET to throw a `System.Net.WebException: Bad value for ai_flags`

Full exception stack trace:
`
System.Net.WebException: Bad value for ai_flags Bad value for ai_flags ---> System.Net.Http.HttpRequestException: Bad value for ai_flags ---> System.Net.Sockets.SocketException: Bad value for ai_flags
  at at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
  --- End of inner exception stack trace ---
  at at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
  at at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
  at at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
  at at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
  at at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
  at at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
  at at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
  at at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
  at at System.Net.HttpWebRequest.SendRequest()
  at at System.Net.HttpWebRequest.GetResponse()
  --- End of inner exception stack trace ---
  at at System.Net.HttpWebRequest.GetResponse()
  at <REDACTED_LOCAL_METHOD>() in <REDACTED_LOCAL_CLASS>
`
